language/vm/src/file_format.rs

```diff
@@ -338,8 +338,37 @@ pub struct FieldInstantiation {
     pub type_parameters: SignatureIndex,
 }
 
-/// A `StructDefinition` is a type definition. It either indicates it is native or
-// defines all the user-specified fields declared on the type.
+/// `Visibility` restricts the accessibility of the associated entity.
+/// - For function visibility, it restricts who may call into the associated function.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+pub enum Visibility {
```

Comment:
I think it is redundant/error-prone to have both this enum and the `u8` constants. I would kill the `enum` and stick to the constants, since that's the style used elsewhere in the file format. The reason it is done this way: using raw bytecodes makes it easier to see what the binary format is, whereas using Rust-specific features like enums can obscure that.

Reply:
Not sure I entirely get this. When speaking about the file format, there may be two things we are talking about:
1. The `CompiledModule`, which is known in the Rust world. In `CompiledModule`, the `Visibility` is an enum,
2. The serialized bytes, which is what is stored on-chain. In the raw bytes, the `Visibility` is a bit in the u8 flag.

A similar situation is the handling of `Kind {All, Resource, Copyable}`. It is an enum in the Rust world, and a `u8` flag in the serialized bytes.

Reply:
I partially agree, and ran into a similar issue with constraints (working on an initial PR for that but was blocked by the versioning thing)

The benefits for the enum here are huge for writing code that checks visibility, granted that code is really rather small. 

I think you could maybe do 
```
#[repr(u8)]
pub enum Visibility {
  Private = 0x0,
  Public = 0x1,
  Protected = 0x2,
}
```
And then use these in the serializer and deserializer. 

So I guess I agree with sam that it is bad to have an enum and the u8 constants, but it is okay if the enum is the same as the u8 constants. It's when you have two things that things go wrong.

Reply:
The flags case is a bit different, but this is how we represent bytecodes (with an repr'd enum and then another bag of opcode constants)

Reply:
Yeah I think deserializing flags into enums is probably fine, especially since they are ensured to be exclusive in the deserializer, putting it not into an enum would feel misleading. 

But I think where `Kind` didn't do the best was having the u8 representation totally different from the enum

Reply:
Got it! Thanks for the description and the suggestion! I think we can go with this:

```
#[repr(u8)]
pub enum Visibility {
  Private = 0x0,
  Public = 0x1,
  Protected = 0x2,
}
```

Reply:
I wouldn't get too eager and switch to that just yet :) I'd like to see if that jives with @sblackshear's concerns

Reply:
I think the enum + `repr` is a great approach. My main concern was indeed about the redundant representation.

Reply:
Updated with 13661c2. Now the bytecode format is a bit more different:

Originally, the `flags` byte has two bits in use:
- 0x1: is_public
- 0x2: is_native

With the new format, the `flags` byte now essentially becomes two bytes:
- byte 1: visibility (with 0 - private, 1 - public, 2 - protected, and in the future, 3 - script, 4 - package, etc...).
- byte 2: now only bit 0x2 is used to indicate whether this is_native.